From 52b21856a09813b53f62157a002ff9d3ad287772 Mon Sep 17 00:00:00 2001
From: Adrian Huang <ahuang12@lenovo.com>
Date: Thu, 27 Feb 2014 11:26:03 +0000
References: FATE#315595
Patch-Mainline: v3.15
Git-Commit: 7969943789df1196faa9ba67518d83fd93e4f9f6
Subject: [PATCH] xhci: add the meaningful IRQ description if it is empty

When some xHCI host controllers fall back to use the legacy IRQ,
the member irq_descr of the usb_hcd structure will be empty. This
leads to the empty string of the xHCI host controller in
/proc/interrupts. Here is the example (The irq 19 is the xHCI host
controller):

           CPU0
  0:         91		IO-APIC-edge      	timer
  8:          1         IO-APIC-edge      	rtc0
  9:       7191         IO-APIC-fasteoi   	acpi
 18:        104       	IR-IO-APIC-fasteoi 	ehci_hcd:usb1, ehci_hcd:usb2
 19:        473     	IR-IO-APIC-fasteoi

After applying the patch, the name of the registered xHCI host
controller can be displayed correctly. Here is the example:

           CPU0
  0:         91		IO-APIC-edge      	timer
  8:          1         IO-APIC-edge      	rtc0
  9:       7191         IO-APIC-fasteoi   	acpi
 18:        104       	IR-IO-APIC-fasteoi 	ehci_hcd:usb1, ehci_hcd:usb2
 19:        473     	IR-IO-APIC-fasteoi	xhci_hcd:usb3

Tested on v3.14-rc4.

Signed-off-by: Adrian Huang <ahuang12@lenovo.com>
Reviewed-by: Nagananda Chumbalkar <nchumbalkar@lenovo.com>
Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 drivers/usb/host/xhci.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index 1a1a755..9cd9bb0 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -386,6 +386,10 @@ static int xhci_try_enable_msi(struct usb_hcd *hcd)
 	}
 
  legacy_irq:
+	if (!strlen(hcd->irq_descr))
+		snprintf(hcd->irq_descr, sizeof(hcd->irq_descr), "%s:usb%d",
+			 hcd->driver->description, hcd->self.busnum);
+
 	/* fall back to legacy interrupt*/
 	ret = request_irq(pdev->irq, &usb_hcd_irq, IRQF_SHARED,
 			hcd->irq_descr, hcd);
-- 
1.8.4.5

